hydromt/models/model_mesh.py

```diff
@@ -253,13 +293,120 @@ def set_mesh(
+
+        # Adding to mesh
         if self._mesh is None:  # NOTE: mesh is initialized with None
             self._mesh = data
         else:
-            for dvar in data.data_vars:
-                if dvar in self._mesh:
-                    self.logger.warning(f"Replacing mesh parameter: {dvar}")
-                self._mesh[dvar] = data[dvar]
+            # Check on crs
+            if not data.ugrid.grid.crs == self.crs:
+                raise ValueError("Data and self.mesh should have the same CRS.")
+            # Save crs as it will be lost when converting to xarray
+            crs = self.crs
+            # Check on new grid topology
+            if grid_name in self.mesh_names:
+                # check if the two grids are the same
+                if (
+                    not self.mesh_grids[grid_name]
+                    .to_dataset()
+                    .equals(data.ugrid.grid.to_dataset())
+                ):
+                    if not overwrite_grid:
+                        raise ValueError(
+                            f"Grid {grid_name} already exists in mesh"
+                            " and has a different topology. "
+                            "Use overwrite_grid=True to overwrite the grid"
+                            " topology and related data."
+                        )
+                    else:
+                        # Remove grid and all corresponding data variables from mesh
+                        self.logger.warning(
+                            f"Overwriting grid {grid_name} and the corresponding"
+                            " data variables in mesh."
+                        )
+                        grids = [
+                            self.mesh_datasets[g].ugrid.to_dataset()
+                            for g in self.mesh_names
+                            if g != grid_name
+                        ]
+                        # Re-define _mesh
+                        grids = xr.merge(grids)
+                        self._mesh = xu.UgridDataset(grids)
+            # Check again mesh_names, could have changed if overwrite_grid=True
+            if grid_name in self.mesh_names:
+                for dvar in data.data_vars:
+                    if dvar in self._mesh:
+                        self.logger.warning(f"Replacing mesh parameter: {dvar}")
+                    self._mesh[dvar] = data[dvar]
+            else:
+                # We are potentially adding a new grid without any data variables
+                self._mesh = xu.UgridDataset(
+                    xr.merge([self.mesh.ugrid.to_dataset(), data.ugrid.to_dataset()])
+                )
+            # Restore crs
+            for grid in self._mesh.ugrid.grids:
+                grid.set_crs(crs)
+
+        # update related geoms if necessary: region
+        if overwrite_grid or new_grid:
+            # add / updates region
+            self._geoms.pop("region", None)
+            self.region
+
+    def get_mesh(
```

Comment:
Reminder to further think about the behaviour of this function. 
Is get_mesh a get method wrapper around mesh_grids and mesh_datasets?
If so, why get_mesh(name), given mesh is None or name does not exist, raises error? 
Instead, if a user calls mesh_grids.get(name) or mesh_grids.get(name) they would be return with none by default. 
Is the idea to harmonize the behaviour or distinct (so that user are forced with use the mesh_grids.get() and meshd_datasets.get())?